### ISSUE TITLE
Implement polynomial algebra

### ISSUE THREAD

**Issue by jobez** [greged93/stark-anatomy#3]
For some reason we don't have tasks lists. So we will do this the less cool way, but I want to make sure theres decent rationale for WHY each of these [building blocks](https://github.com/aszepieniec/stark-anatomy/blob/master/docs/basic-tools.md?plain=1#L1) are being summoned

<img src="https://aszepieniec.github.io/stark-anatomy/graphics/pipeline.svg"></img>

**Setup phase:**

- Finite field arithmetic - Construct finite field F and subgroups G1, G2 needed for STARK 

**Prover:**

- Univariate polynomials:
  - Interpolation - Compute trace function T(x) 
  - Evaluation - Evaluate T(x) on random subset to get trace table
- Multivariate polynomials:
  - Symbolic evaluation - Evaluate constraint polynomials on trace function to get (univariate) quotient polynomial

**Verifier:** 

- Univariate polynomials:
  - Interpolation - Interpolate quotient polynomial
  - Evaluation - Evaluate interpolated polynomial on random subset to check trace table
- Multivariate polynomials:
  - Lifting - Lift interpolated univariate polynomial back to multivariate form  
  - Evaluation - Evaluate lifted polynomial symbolically on trace table points to reconstruct quotient polynomial

**Common:**

- Finite field arithmetic - Underlies all polynomial operations
- Arithmetic operators - Used throughout for operating on polynomials
- Additional ops (e.g. division) - Used as needed in prover and verifier steps

In summary:
- Univariate polynomial ops (interp, eval, scaling, etc) are used in trace function computation
- Multivariate polynomial ops (symbolic eval, lifting) are used for constraint polynomials 
- All polynomial ops rely on underlying finite field arithmetic.

I will breakdown issues for univariate and multivariate polynomials
